grayscale.py

```diff
@@ -1,336 +1,1055 @@
+                idle()
+        self.cs(1)
+        self.reset()
+        self.cs(0)
+        self.dc(0)
+        # reinitialise to the normal configuration
+        self.spi.write(bytearray([
+            0xae, 0x20,0x00, 0x40, 0xa1, 0xa8,self.height-1, 0xc8, 0xd3,0, 0xda,0x12, 0xd5,0x80,
+            0xd9,0xf1, 0xdb,0x20, 0x81,0x7f,
+            0xa4, 0xa6, 0x8d,0x14, 0xad,0x30, 0xaf,
+            0x21,28,99, 0x22,0,4]))
+        self.cs(1)
 
-    def saveConfig(self):
-        ConfigFile.save(self.config)
 
-    ### Be a proxy to GraphicsClass
+    def pause(self):
+        if self.error:
+            return False
+        if self._state[Grayscale_StateIndex_State] == Grayscale_ThreadState_Paused:
+            return False
+        self._state[Grayscale_StateIndex_State] = Grayscale_ThreadState_Pausing
+        while self._state[Grayscale_StateIndex_State] != Grayscale_ThreadState_Paused:
+            idle()
+        return True
+
+    def resume(self):
+        if self._state[Grayscale_StateIndex_State] != Grayscale_ThreadState_Paused:
+            return
+        self._state[Grayscale_StateIndex_State] = Grayscale_ThreadState_Resuming
+        while self._state[Grayscale_StateIndex_State] != Grayscale_ThreadState_Running:
+            idle()
+
+    def write_cmd(self, cmd):
+        if cmd is list:
+            cmd = bytearray(cmd)
+        elif not cmd is bytearray:
+            cmd = bytearray([cmd])
+        if self._state[Grayscale_StateIndex_State] == Grayscale_ThreadState_Running:
+            self.pending_cmds = cmd
+            self._state[Grayscale_StateIndex_PendingCmd] = 1
+            while self._state[Grayscale_StateIndex_PendingCmd]:
+                idle()
+            self.pending_cmds = None
+        else:
+            self.dc(0)
+            self.spi.write(cmd)
+
+    def poweroff(self):
+        self.write_cmd(0xae)
+    def poweron(self):
+        self.write_cmd(0xaf)
 
     @micropython.native
-    def setFont(self, fontFile, width, height, space):
-        self.gsGraphics1.setFont(fontFile, width, height, space)
-        self.gsGraphics2.setFont(fontFile, width, height, space)
+    def show(self):
+        self._state[Grayscale_StateIndex_CopyBuffs] = 1
+        if self._state[Grayscale_StateIndex_State] != Grayscale_ThreadState_Running:
+            return
+        while self._state[Grayscale_StateIndex_CopyBuffs] != 0:
+            idle()
+
+    @micropython.native
+    def show_async(self):
+        self._state[Grayscale_StateIndex_CopyBuffs] = 1
+
 
     @micropython.native
-    def setFPS(self):
-        pass
+    def setFPS(self, newFrameRate):
+        self.frameRate = newFrameRate
+        if newFrameRate != 0:
+            self.frameTimeMs = 1000 // newFrameRate
 
     @micropython.native
     def update(self):
-        pass
+        self.show()
+        if self.frameRate > 0:
+            frameTimeMs = self.frameTimeMs
+            lastUpdateEnd = self.lastUpdateEnd
```

Comment:
I don't think this variable is ever used

Reply:
It should be! I forgot to change the following `self.lastUpdateEnd` references to just `lastUpdateEnd`.